File: src/mapping/script/lidar.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import rospy
import numpy
from sensor_msgs.msg import PointCloud2
from sensor_msgs import point_cloud2
from sensor_msgs.msg import PointField

# 读取雷达信息回调函数
def lidar_Callback(lidar_msg):
    lidar_data = point_cloud2.read_points_list(lidar_msg)
    # print(type(lidar_data))
    # for p in lidar_data:
    #     print(p)

    lidar_data = numpy.array(lidar_data)
    # output.header.stamp = rospy.Time().now()
    # output.height = 1
    # output.width = lidar_msg.width
    # output.fields = [
    #     PointField('x', 0, PointField.FLOAT32, 1),
    #     PointField('y', 4, PointField.FLOAT32, 1),
    #     PointField('z', 8, PointField.FLOAT32, 1)]
    # output.is_bigendian = False
    # output.point_step = 12
    # output.is_dense = False
    output.data = numpy.asarray(lidar_data, numpy.float32).tostring()
    lidar_pub.publish(output)

if __name__ == '__main__':
    '''-------------------------------程序初始化部分------------------------------------'''
    # ROS节点初始化
    rospy.init_node('lidar', anonymous=True)

    rospy.Subscriber("/airsim_node/drone_1/lidar/LidarSensor", PointCloud2, lidar_Callback)

    # 发布点云信息
    lidar_pub = rospy.Publisher('/lidar_test', PointCloud2, queue_size=10)

    rate = rospy.Rate(10)

    output = PointCloud2()
    output.header.frame_id = "world_enu"; 
    '''-------------------------------程序循环部分------------------------------------'''
    # while not rospy.is_shutdown():
    #     output.header.stamp = rospy.Time().now()


    rospy.spin()

File: src/mapping/src/lidar.cpp
#include <ros/ros.h>
#include <sensor_msgs/PointCloud2.h>
#include <pcl/point_cloud.h>
#include <pcl_conversions/pcl_conversions.h> 
#include <pcl_ros/transforms.h>

ros::Publisher pcl_pub;
// 实例化对象  用于存放可以发布的点云形式，
// sensor_msgs::PointCloud2 output; 
// 实例化对象  用于存放PCL点云数据
// pcl::PointCloud<pcl::PointXYZ> cloud; 

void lidar_Callback(const sensor_msgs::PointCloud2::ConstPtr & lidar_msg)
{
    // ROS_INFO("lidar data width is:%d", lidar_msg -> width);
    // ROS_INFO("lidar data size is:%ld", sizeof(lidar_msg -> data));
    // ROS_INFO("lidar data size1 is:%ld", sizeof(lidar_msg -> data.size()));
	// pcl::PointCloud<pcl::PointXYZ> output;

	pcl::PointCloud<pcl::PointXYZ>::Ptr output(new pcl::PointCloud<pcl::PointXYZ>);
	pcl::fromROSMsg(*lidar_msg, *output);
	
	// output.header.frame_id = "world_enu"; 
	// output->header.frame_id = "world_enu"; 
	// pcl_pub.publish(output);
}

int main (int argc, char **argv) 
{
    ros::init (argc, argv, "pcl_create"); 
	
	ros::NodeHandle nh; 
	// 创建发布者
	pcl_pub = nh.advertise<sensor_msgs::PointCloud2> ("/lidar/fly", 1);

    ros::Subscriber person_info_sub = nh.subscribe("/airsim_node/drone_1/lidar/LidarSensor", 10, lidar_Callback);

    // 等待循环
	ros::spin();

    return 0;
}